plugins/notifications-backend/src/service/router.ts

```diff
@@ -470,71 +485,84 @@ export async function createRouter(
-        user,
-      };
-      const notification = await preProcessNotification(userNotification, opts);
+    scope?: string,
+  ): Promise<Notification | undefined> => {
+    const userNotification = {
+      ...baseNotification,
+      id: uuid(),
+      user,
+    };
+    const notification = await preProcessNotification(userNotification, opts);
+
+    const enabled = await isNotificationsEnabled({
+      user,
+      channel: WEB_NOTIFICATION_CHANNEL,
+      origin: userNotification.origin,
+    });
+
+    let ret = notification;
+
+    if (!enabled) {
+      postProcessNotification(ret, opts);
+      return undefined;
+    }
 
-      const enabled = await isNotificationsEnabled({
+    let existingNotification;
+    if (scope) {
+      existingNotification = await store.getExistingScopeNotification({
         user,
-        channel: WEB_NOTIFICATION_CHANNEL,
-        origin: userNotification.origin,
+        scope,
+        origin,
       });
+    }
 
-      let ret = notification;
-      if (enabled) {
-        let existingNotification;
-        if (scope) {
-          existingNotification = await store.getExistingScopeNotification({
-            user,
-            scope,
-            origin,
-          });
-        }
-
-        if (existingNotification) {
-          const restored = await store.restoreExistingNotification({
-            id: existingNotification.id,
-            notification,
-          });
-          ret = restored ?? notification;
-        } else {
-          await store.saveNotification(notification);
-        }
+    if (existingNotification) {
+      const restored = await store.restoreExistingNotification({
+        id: existingNotification.id,
+        notification,
+      });
+      ret = restored ?? notification;
+    } else {
+      await store.saveNotification(notification);
+    }
 
-        notifications.push(ret);
-
-        if (signals) {
-          await signals.publish<NewNotificationSignal>({
-            recipients: { type: 'user', entityRef: [user] },
-            message: {
-              action: 'new_notification',
-              notification_id: ret.id,
-            },
-            channel: 'notifications',
-          });
-        }
-      }
-      postProcessNotification(ret, opts);
+    if (signals) {
+      await signals.publish<NewNotificationSignal>({
+        recipients: { type: 'user', entityRef: [user] },
+        message: {
+          action: 'new_notification',
+          notification_id: ret.id,
+        },
+        channel: 'notifications',
+      });
     }
-    return notifications;
+    postProcessNotification(ret, opts);
+    return ret;
+  };
+
+  const sendUserNotifications = async (
+    baseNotification: Omit<Notification, 'user' | 'id'>,
+    users: string[],
+    opts: NotificationSendOptions,
+    origin: string,
+  ): Promise<Notification[]> => {
+    const { scope } = opts.payload;
+    const uniqueUsers = [...new Set(users)];
+    const throttled = throttle((user: string) =>
```

Comment:
Just noting here that this is throttled "per send", not globally. So several large overlapping sends still stack up. Maybe that's OK, just noting.

Reply:
Yep, it's per send 👍